File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
N = int(input())

maps = []
result = []

# 지도 입력
for _ in range(N):
    maps.append(input())

# 방문 여부 기록
visited = [[0] * N for _ in range(N)]

# 상하좌우 방향 설정
direction = [(-1, 0), (1, 0), (0, -1), (0, 1)]

# DFS 정의
def dfs(x, y):
    global count
    for dx, dy in direction:
        nx, ny = x + dx, y + dy
        if 0 <= nx < N and 0 <= ny < N:
            if maps[nx][ny] == '1' and visited[nx][ny] == 0:
                visited[nx][ny] = 1
                count += 1
                dfs(nx, ny)

# 전체 지도 순회
for i in range(N):
    for j in range(N):
        if maps[i][j] == '1' and visited[i][j] == 0:
            count = 1
            visited[i][j] = 1  # 시작점 방문 체크!
            dfs(i, j)
            result.append(count)

# 출력
print(len(result))         # 총 단지 수
for val in sorted(result): # 단지 내 집 수 오름차순 출력
    print(val)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;


int N;
int cnt;
vector<string> maps;
vector<int> result;
vector<vector<int>> visited;
vector<pair<int, int>> direction = {{0,1}, {1,0}, {-1,0}, {0,-1}};

void dfs(int x, int y) {

    for (auto dir: direction) {
        int nx = x + dir.first;
        int ny = y + dir.second;

        if (0 <= nx && nx < N && 0 <= ny && ny < N && maps[nx][ny] == '1' && visited[nx][ny] == 0)
        {
            visited[nx][ny] = 1;
            dfs(nx, ny);
            cnt++;
        }
    }

}


int main() {
    cin >> N;
    visited.resize(N, vector<int>(N,0));

    for (int i=0; i<N; i++) {
        string a;
        cin >> a;
        maps.push_back(a); 
    }

    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            
            if (visited[i][j] == 0 && maps[i][j] == '1') {
                cnt = 1;
                visited[i][j] = 1;
                dfs(i, j);
            result.push_back(cnt);
            }
        }
    }

    sort(result.begin(), result.end());

    cout << result.size() << '\n';

    for (int i=0; i<result.size(); i++) {
        cout << result[i] << '\n';
    }


}